scripts/test_solidity_lsp.py

```diff
@@ -0,0 +1,624 @@
+    def __init__(self, msg: str):
+        super().__init__("Bad header: " + msg)
+
+class MyEncoder(json.JSONEncoder):
+    """
+    Encodes an object in JSON
+    """
+    def default(self, o):
+        return o.__dict__
+
+class JsonRpcProcess:
+    exe_path: str
+    exe_args: List[str]
+    process: subprocess.Popen
+    trace_io: bool
+
+    def __init__(self, exe_path: str, exe_args: List[str], trace_io: bool = True):
+        self.exe_path = exe_path
+        self.exe_args = exe_args
+        self.trace_io = trace_io
+
+    def __enter__(self):
+        self.process = subprocess.Popen(
+            [self.exe_path, *self.exe_args],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE
+        )
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback) -> None:
+        self.process.kill()
+        self.process.wait(timeout=2.0)
+
+    def trace(self, topic: str, message: str) -> None:
+        if self.trace_io:
+            print(f"{SGR_TRACE}{topic}:{SGR_RESET} {message}")
+
+    def receive_message(self) -> Union[None, dict]:
+        # Note, we should make use of timeout to avoid infinite blocking if nothing is received.
+        CONTENT_LENGTH_HEADER = "Content-Length: "
+        CONTENT_TYPE_HEADER = "Content-Type: "
+        if self.process.stdout == None:
+            return None
+        message_size = None
+        while True:
+            # read header
+            line = self.process.stdout.readline()
+            if line == '':
+                # server quit
+                return None
+            line = line.decode("utf-8")
+            if not line.endswith("\r\n"):
+                raise BadHeader("missing newline")
+            # remove the "\r\n"
+            line = line[:-2]
+            if line == "":
+                break # done with the headers
+            if line.startswith(CONTENT_LENGTH_HEADER):
+                line = line[len(CONTENT_LENGTH_HEADER):]
+                if not line.isdigit():
+                    raise BadHeader("size is not int")
+                message_size = int(line)
+            elif line.startswith(CONTENT_TYPE_HEADER):
+                # nothing todo with type for now.
+                pass
+            else:
+                raise BadHeader("unknown header")
+        if message_size is None:
+            raise BadHeader("missing size")
+        rpc_message = self.process.stdout.read(message_size).decode("utf-8")
+        json_object = json.loads(rpc_message)
+        self.trace('receive_message', json.dumps(json_object, indent=4, sort_keys=True))
+        return json_object
+
+    def send_message(self, method_name: str, params: Optional[dict]) -> None:
+        if self.process.stdin == None:
+            return
+        message = {
+            'jsonrpc': '2.0',
+            'method': method_name,
+            'params': params
+        }
+        json_string = json.dumps(obj=message, cls=MyEncoder)
```

Comment:
I think @cameel alraedy asked that - do we really need MyEncoder?

Reply:
No, I've put that in because a few of his proposals actually broke the tests again. So a tried to revert. I'll drop this part, we in fact don't need it (tested already).